common-rule-engine/src/main/java/org/sonar/commonruleengine/Engine.java

```
@@ -20,25 +20,38 @@
 package org.sonar.commonruleengine;
 
 import java.io.IOException;
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
 import java.util.Set;
 import java.util.stream.Collectors;
 import org.sonar.api.batch.fs.InputFile;
 import org.sonar.commonruleengine.checks.Check;
 import org.sonar.uast.UastNode;
+import org.sonar.uast.validators.Validator;
+import org.sonar.uast.validators.ValidatorList;
 
 public class Engine {
 
   private final EngineContext engineContext;
   private final MetricsVisitor metricsVisitor;
 
   public Engine(Collection<Check> rules) {
+    this(rules, ValidatorList.all());
+  }
+
+  public Engine(Collection<Check> rules, Collection<Validator> validators) {
     engineContext = new EngineContext();
     metricsVisitor = new MetricsVisitor();
-    rules.forEach(rule -> rule.initialize(engineContext));
+    // should this be parameterized in order to switch ON/OFF validation?
+    Collection<Check> rulesWithValidators = new ArrayList<>();
+    // add validators first, to validate before playing any rule
+    rulesWithValidators.addAll(validators);
```

Question I see here is  : shall we fail the analysis or "skip" the invalid uast for that file and log a big error if it is wrong. Let's ask this to PM @agigleux .

---

Currently an exception is going to be thrown, meaning that no rule is  going to be played against it, and an explicit debug message is going to be logged. See GoSensor file